Add initial support for _WKWebExtensionWindow and _WKWebExtensionTab APIs to _WKWebExtensionController and _WkWebExtensionContext.

xeenon opened the pull request:
#### fd7774b7dda952470cd5dfed81eb97f08ca8cdbb
<pre>
Add initial support for _WKWebExtensionWindow and _WKWebExtensionTab APIs to _WKWebExtensionController and _WkWebExtensionContext.
<a href="https://bugs.webkit.org/show_bug.cgi?id=260156">https://bugs.webkit.org/show_bug.cgi?id=260156</a>

Reviewed by Brian Weinstein.

Implements the main window and tab management APIs on _WKWebExtensionController and _WkWebExtensionContext for tracking open windows,
tabs and the focused window. Also corrects some missing nullability attributes on some methods and properties.

Added new WKWebExtensionWindow and WKWebExtensionTab tests for some of these APIs.

* Source/WTF/wtf/WeakObjCPtr.h:
(WTF::WeakObjCPtr::operator! const): Made one line.
(WTF::WeakObjCPtr::operator bool const): Added.
* Source/WebKit/Platform/cocoa/CocoaHelpers.h: Added THROW_UNLESS macro.
* Source/WebKit/Shared/Extensions/WebExtensionTabIdentifier.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionWindowIdentifier.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext openWindows]): Added.
(-[_WKWebExtensionContext focusedWindow]): Added.
(-[_WKWebExtensionContext openTabs]): Added.
(-[_WKWebExtensionContext didOpenWindow:]): Added.
(-[_WKWebExtensionContext didCloseWindow:]): Added.
(-[_WKWebExtensionContext didFocusWindow:]): Added.
(-[_WKWebExtensionContext didOpenTab:]): Added.
(-[_WKWebExtensionContext didCloseTab:windowIsClosing:]): Added.
(-[_WKWebExtensionContext didSelectTabs:]): Added.
(-[_WKWebExtensionContext didMoveTab:fromIndex:inWindow:]): Added.
(-[_WKWebExtensionContext didReplaceTab:withTab:]): Added.
(-[_WKWebExtensionContext didChangeTabProperties:forTab:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController extensions]): Stop using a variable here.
(-[_WKWebExtensionController didOpenWindow:]): Added.
(-[_WKWebExtensionController didCloseWindow:]): Added.
(-[_WKWebExtensionController didFocusWindow:]): Added.
(-[_WKWebExtensionController didOpenTab:]): Added.
(-[_WKWebExtensionController didCloseTab:windowIsClosing:]): Added.
(-[_WKWebExtensionController didSelectTabs:]): Added.
(-[_WKWebExtensionController didMoveTab:fromIndex:inWindow:]): Added.
(-[_WKWebExtensionController didReplaceTab:withTab:]): Added.
(-[_WKWebExtensionController didChangeTabProperties:forTab:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerDelegate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionTab.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindow.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITestCocoa.mm:
(WebKit::WebExtensionContext::testResult): Use new delegate getter for cleaner approch.
(WebKit::WebExtensionContext::testEqual): Ditto.
(WebKit::WebExtensionContext::testMessage): Ditto.
(WebKit::WebExtensionContext::testYielded): Ditto.
(WebKit::WebExtensionContext::testFinished): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load): Call populateWindowsAndTabs().
(WebKit::WebExtensionContext::getOrCreateWindow): Added.
(WebKit::WebExtensionContext::getWindow): Added.
(WebKit::WebExtensionContext::getOrCreateTab): Added.
(WebKit::WebExtensionContext::getTab): Added.
(WebKit::WebExtensionContext::populateWindowsAndTabs): Added.
(WebKit::WebExtensionContext::openWindows): Added.
(WebKit::WebExtensionContext::openTabs): Added.
(WebKit::WebExtensionContext::focusedWindow): Added.
(WebKit::WebExtensionContext::didOpenWindow): Added.
(WebKit::WebExtensionContext::didCloseWindow): Added.
(WebKit::WebExtensionContext::didFocusWindow): Added.
(WebKit::WebExtensionContext::didOpenTab): Added.
(WebKit::WebExtensionContext::didCloseTab): Added.
(WebKit::WebExtensionContext::didSelectTab): Added.
(WebKit::WebExtensionContext::didMoveTab): Added.
(WebKit::WebExtensionContext::didReplaceTab): Added.
(WebKit::WebExtensionContext::didChangeTabProperties): Added.
(WebKit::WebExtensionContext::webViewConfiguration):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm: Added.
(WebKit::WebExtensionTab::WebExtensionTab):
(WebKit::WebExtensionTab::window const):
(WebKit::WebExtensionTab::parentTab const):
(WebKit::WebExtensionTab::mainWebView const):
(WebKit::WebExtensionTab::webViews const):
(WebKit::WebExtensionTab::title const):
(WebKit::WebExtensionTab::isSelected const):
(WebKit::WebExtensionTab::isPinned const):
(WebKit::WebExtensionTab::toggleReaderMode):
(WebKit::WebExtensionTab::isReaderModeAvailable const):
(WebKit::WebExtensionTab::isShowingReaderMode const):
(WebKit::WebExtensionTab::mute):
(WebKit::WebExtensionTab::unmute):
(WebKit::WebExtensionTab::isAudible const):
(WebKit::WebExtensionTab::isMuted const):
(WebKit::WebExtensionTab::size const):
(WebKit::WebExtensionTab::zoomFactor const):
(WebKit::WebExtensionTab::url const):
(WebKit::WebExtensionTab::pendingURL const):
(WebKit::WebExtensionTab::isLoadingComplete const):
(WebKit::WebExtensionTab::detectWebpageLocale):
(WebKit::WebExtensionTab::loadURL):
(WebKit::WebExtensionTab::reload):
(WebKit::WebExtensionTab::reloadFromOrigin):
(WebKit::WebExtensionTab::goBack):
(WebKit::WebExtensionTab::goForward):
(WebKit::WebExtensionTab::close):
(WebKit::WebExtensionTab::select):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionWindowCocoa.mm: Added.
(WebKit::WebExtensionWindow::WebExtensionWindow):
(WebKit::WebExtensionWindow::operator== const):
(WebKit::WebExtensionWindow::tabs const):
(WebKit::WebExtensionWindow::activeTab const):
(WebKit::toImpl):
(WebKit::WebExtensionWindow::type const):
(WebKit::WebExtensionWindow::state const):
(WebKit::WebExtensionWindow::isFocused const):
(WebKit::WebExtensionWindow::isEphemeral const):
(WebKit::WebExtensionWindow::frame const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::didChangeTabProperties):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::delegate const):
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h: Added.
(WebKit::WebExtensionTab::create):
(WebKit::WebExtensionTab::identifier const):
(WebKit::WebExtensionTab::extensionContext const):
(WebKit::WebExtensionTab::delegate const):
(WebKit::WebExtensionTab::isValid const):
* Source/WebKit/UIProcess/Extensions/WebExtensionWindow.h: Added.
(WebKit::WebExtensionWindow::create):
(WebKit::WebExtensionWindow::identifier const):
(WebKit::WebExtensionWindow::extensionContext const):
(WebKit::WebExtensionWindow::delegate const):
(WebKit::WebExtensionWindow::isValid const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionTab.mm: Added.
(-[TestWindowForTabs tabsForWebExtensionContext:]):
(-[TestWindowForTabs activeTabForWebExtensionContext:]):
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionWindow.mm: Added.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm: Fix build by including missing headers.
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate webExtensionController:openWindowsForExtensionContext:]):
(-[TestWebExtensionsDelegate webExtensionController:focusedWindowForExtensionContext:]):

Canonical link: <a href="https://commits.webkit.org/267104@main">https://commits.webkit.org/267104@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9b8a8b01d2f34657cbebc45f612c6cbde7e32d6b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15671 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15978 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16351 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17428 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14706 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15853 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18948 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16080 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15856 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16332 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13343 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18174 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13593 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14145 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21064 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/13465 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14594 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14311 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17570 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/14914 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14903 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12636 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/15862 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14157 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3990 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18521 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/16101 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1910 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14722 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3799 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

davidjohnson91 commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
```diff
@@ -453,6 +453,118 @@ WK_CLASS_AVAILABLE(macos(13.3), ios(16.4))
+ @discussion This method informs only the specific extension that a tab has been selected. If the intention is to inform all loaded
+ extensions consistently, you should use the respective method on the extension controller instead.
+ */
+- (void)didSelectTab:(nullable id <_WKWebExtensionTab>)selectedTab;
```
We should consider how multi-select of tabs works with this API.

xeenon replied:
Changed this to `didSelectTabs:` and takes a `NSSet`.

b-weinstein commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
```diff
@@ -453,6 +453,119 @@ WK_CLASS_AVAILABLE(macos(13.3), ios(16.4))
+ @discussion This method informs only the specific extension that a tab has been moved. If the intention is to inform all loaded 
+ extensions consistently, you should use the respective method on the extension controller instead.
+ */
+- (void)didMoveTab:(id <_WKWebExtensionTab>)movedTab fromIndex:(NSUInteger)index inWindow:(nullable id <_WKWebExtensionWindow>)oldWindow NS_SWIFT_NAME(didMoveTab(_:from:in:));
```
How will we get the new index of the tab in the window?

xeenon replied:
Will ask the `movedTab`'s `window`'s `tabs` array for the current index of the object. It should be moved in the array before this call.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1131,6 +1138,271 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    }
+
+    if (!tab->isValid()) {
+        RELEASE_LOG_ERROR(Extensions, "Tab with identifier %{public}llu has nil delegate; reference not removed via didCloseWindow: before release", identifier.toUInt64());
```
Should this be `didCloseTab:`?

xeenon replied:
Yep!

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1131,6 +1138,271 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    ASSERT(m_windowMap.contains(window.identifier()));
+    ASSERT(m_windowMap.get(window.identifier()) == &window);
+
+    m_focusedWindowIdentifier = window.identifier();
```
I guess it's not possible to make a window that isn't focused at first, is it?

xeenon replied:
Right, we assume any newly opened window is the focused one. But maybe that isn't the right assumption. 🤔

I might remove this.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1131,6 +1138,271 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    if (!isLoaded())
+        return;
+
+    // FIXME: Fire event here.
```
These could use a bug number?

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
```diff
@@ -0,0 +1,366 @@
+    if (m_respondsToWebViews) {
+        auto *webViews = [m_delegate webViewsForWebExtensionContext:m_extensionContext->wrapper()];
+        THROW_UNLESS([webViews isKindOfClass:NSArray.class], @"Object returned by webViewsForWebExtensionContext: is not an array");
+        THROW_UNLESS([webViews containsObject:mainWebView], @"Array returned by webViewsForWebExtensionContext: does not contain the main web view");
```
Do we want to check everything in the array is a WKWebView?

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
```diff
@@ -0,0 +1,366 @@
+    for (WKWebView *webView in webViews) {
+        THROW_UNLESS([webView isKindOfClass:WKWebView.class], @"Object in array returned by webViewsForWebExtensionContext: is not a WKWebView");
+        THROW_UNLESS(webView.configuration._webExtensionController == extensionContext()->extensionController()->wrapper(), @"WKWebView returned by webViewsForWebExtensionContext: is not configured with the same _WKWebExtensionController as extension context");
+    }
```
Oh, you do that here! Never mind.

xeenon committed:
Add initial support for _WKWebExtensionWindow and _WKWebExtensionTab APIs to _WKWebExtensionController and _WkWebExtensionContext.
https://bugs.webkit.org/show_bug.cgi?id=260156

Reviewed by Brian Weinstein.

Implements the main window and tab management APIs on _WKWebExtensionController and _WkWebExtensionContext for tracking open windows,
tabs and the focused window. Also corrects some missing nullability attributes on some methods and properties.

Added new WKWebExtensionWindow and WKWebExtensionTab tests for some of these APIs.

* Source/WTF/wtf/WeakObjCPtr.h:
(WTF::WeakObjCPtr::operator! const): Made one line.
(WTF::WeakObjCPtr::operator bool const): Added.
* Source/WebKit/Platform/cocoa/CocoaHelpers.h: Added THROW_UNLESS macro.
* Source/WebKit/Shared/Extensions/WebExtensionTabIdentifier.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionWindowIdentifier.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext openWindows]): Added.
(-[_WKWebExtensionContext focusedWindow]): Added.
(-[_WKWebExtensionContext openTabs]): Added.
(-[_WKWebExtensionContext didOpenWindow:]): Added.
(-[_WKWebExtensionContext didCloseWindow:]): Added.
(-[_WKWebExtensionContext didFocusWindow:]): Added.
(-[_WKWebExtensionContext didOpenTab:]): Added.
(-[_WKWebExtensionContext didCloseTab:windowIsClosing:]): Added.
(-[_WKWebExtensionContext didSelectTabs:]): Added.
(-[_WKWebExtensionContext didMoveTab:fromIndex:inWindow:]): Added.
(-[_WKWebExtensionContext didReplaceTab:withTab:]): Added.
(-[_WKWebExtensionContext didChangeTabProperties:forTab:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController extensions]): Stop using a variable here.
(-[_WKWebExtensionController didOpenWindow:]): Added.
(-[_WKWebExtensionController didCloseWindow:]): Added.
(-[_WKWebExtensionController didFocusWindow:]): Added.
(-[_WKWebExtensionController didOpenTab:]): Added.
(-[_WKWebExtensionController didCloseTab:windowIsClosing:]): Added.
(-[_WKWebExtensionController didSelectTabs:]): Added.
(-[_WKWebExtensionController didMoveTab:fromIndex:inWindow:]): Added.
(-[_WKWebExtensionController didReplaceTab:withTab:]): Added.
(-[_WKWebExtensionController didChangeTabProperties:forTab:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerDelegate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionTab.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindow.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPITestCocoa.mm:
(WebKit::WebExtensionContext::testResult): Use new delegate getter for cleaner approch.
(WebKit::WebExtensionContext::testEqual): Ditto.
(WebKit::WebExtensionContext::testMessage): Ditto.
(WebKit::WebExtensionContext::testYielded): Ditto.
(WebKit::WebExtensionContext::testFinished): Ditto.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load): Call populateWindowsAndTabs().
(WebKit::WebExtensionContext::getOrCreateWindow): Added.
(WebKit::WebExtensionContext::getWindow): Added.
(WebKit::WebExtensionContext::getOrCreateTab): Added.
(WebKit::WebExtensionContext::getTab): Added.
(WebKit::WebExtensionContext::populateWindowsAndTabs): Added.
(WebKit::WebExtensionContext::openWindows): Added.
(WebKit::WebExtensionContext::openTabs): Added.
(WebKit::WebExtensionContext::focusedWindow): Added.
(WebKit::WebExtensionContext::didOpenWindow): Added.
(WebKit::WebExtensionContext::didCloseWindow): Added.
(WebKit::WebExtensionContext::didFocusWindow): Added.
(WebKit::WebExtensionContext::didOpenTab): Added.
(WebKit::WebExtensionContext::didCloseTab): Added.
(WebKit::WebExtensionContext::didSelectTab): Added.
(WebKit::WebExtensionContext::didMoveTab): Added.
(WebKit::WebExtensionContext::didReplaceTab): Added.
(WebKit::WebExtensionContext::didChangeTabProperties): Added.
(WebKit::WebExtensionContext::webViewConfiguration):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm: Added.
(WebKit::WebExtensionTab::WebExtensionTab):
(WebKit::WebExtensionTab::window const):
(WebKit::WebExtensionTab::parentTab const):
(WebKit::WebExtensionTab::mainWebView const):
(WebKit::WebExtensionTab::webViews const):
(WebKit::WebExtensionTab::title const):
(WebKit::WebExtensionTab::isSelected const):
(WebKit::WebExtensionTab::isPinned const):
(WebKit::WebExtensionTab::toggleReaderMode):
(WebKit::WebExtensionTab::isReaderModeAvailable const):
(WebKit::WebExtensionTab::isShowingReaderMode const):
(WebKit::WebExtensionTab::mute):
(WebKit::WebExtensionTab::unmute):
(WebKit::WebExtensionTab::isAudible const):
(WebKit::WebExtensionTab::isMuted const):
(WebKit::WebExtensionTab::size const):
(WebKit::WebExtensionTab::zoomFactor const):
(WebKit::WebExtensionTab::url const):
(WebKit::WebExtensionTab::pendingURL const):
(WebKit::WebExtensionTab::isLoadingComplete const):
(WebKit::WebExtensionTab::detectWebpageLocale):
(WebKit::WebExtensionTab::loadURL):
(WebKit::WebExtensionTab::reload):
(WebKit::WebExtensionTab::reloadFromOrigin):
(WebKit::WebExtensionTab::goBack):
(WebKit::WebExtensionTab::goForward):
(WebKit::WebExtensionTab::close):
(WebKit::WebExtensionTab::select):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionWindowCocoa.mm: Added.
(WebKit::WebExtensionWindow::WebExtensionWindow):
(WebKit::WebExtensionWindow::operator== const):
(WebKit::WebExtensionWindow::tabs const):
(WebKit::WebExtensionWindow::activeTab const):
(WebKit::toImpl):
(WebKit::WebExtensionWindow::type const):
(WebKit::WebExtensionWindow::state const):
(WebKit::WebExtensionWindow::isFocused const):
(WebKit::WebExtensionWindow::isEphemeral const):
(WebKit::WebExtensionWindow::frame const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::didChangeTabProperties):
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::delegate const):
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h: Added.
(WebKit::WebExtensionTab::create):
(WebKit::WebExtensionTab::identifier const):
(WebKit::WebExtensionTab::extensionContext const):
(WebKit::WebExtensionTab::delegate const):
(WebKit::WebExtensionTab::isValid const):
* Source/WebKit/UIProcess/Extensions/WebExtensionWindow.h: Added.
(WebKit::WebExtensionWindow::create):
(WebKit::WebExtensionWindow::identifier const):
(WebKit::WebExtensionWindow::extensionContext const):
(WebKit::WebExtensionWindow::delegate const):
(WebKit::WebExtensionWindow::isValid const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionTab.mm: Added.
(-[TestWindowForTabs tabsForWebExtensionContext:]):
(-[TestWindowForTabs activeTabForWebExtensionContext:]):
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionWindow.mm: Added.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebsitePolicies.mm: Fix build by including missing headers.
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.h:
* Tools/TestWebKitAPI/cocoa/TestWebExtensionsDelegate.mm:
(-[TestWebExtensionsDelegate webExtensionController:openWindowsForExtensionContext:]):
(-[TestWebExtensionsDelegate webExtensionController:focusedWindowForExtensionContext:]):

Canonical link: https://commits.webkit.org/267104@main
